server/src/main/java/org/elasticsearch/snapshots/RestoreService.java

```diff
@@ -388,21 +388,31 @@ private void startRestore(
         // Remove the data streams from the list of requested indices
         requestIndices.removeAll(dataStreamsToRestore.keySet());
 
-        // And add the backing indices
-        final Set<String> nonSystemDataStreamIndices;
-        final Set<String> systemDataStreamIndices;
+        // And add the backing indices and failure indices of data streams (the distinction is important for renaming)
+        final DataStreamIndices nonSystemDataStreamIndices;
+        final DataStreamIndices systemDataStreamIndices;
         {
```

Comment:
I tried coming up with an alternative to using a record, as I wasn't sure how useful it was in this scenario. What about doing something like:
```java
        final List<String> systemDataStreamIndices = new ArrayList<>();
        final List<String> nonSystemDataStreamBackingIndices = new ArrayList<>();
        final List<String> nonSystemDataStreamFailureIndices = new ArrayList<>();
        {
            for (DataStream ds : dataStreamsToRestore.values()) {
                for (Index idx : ds.getIndices()) {
                    if (ds.isSystem()) {
                        systemDataStreamIndices.add(idx.getName());
                    } else {
                        nonSystemDataStreamBackingIndices.add(idx.getName());
                    }
                }
                for (Index idx : ds.getFailureIndices()) {
                    if (ds.isSystem()) {
                        systemDataStreamIndices.add(idx.getName());
                    } else {
                        nonSystemDataStreamFailureIndices.add(idx.getName());
                    }
                }
            }
        }
```
And then passing `nonSystemDataStream(Backing|Failure)Indices` as separate arguments to the rename function?

Reply:
I'm also a bit averse to the record usage here. It's easier to pass the indices down to the methods that need them, but they'll still need to be unpacked once they get there. I know we discussed that there might be a time when we have other backing indices on a data stream, but I think this code would be easy enough to update in that hypothetical use case.

Reply:
Removed